packages/devtools_app/lib/src/debugger/flutter/console.dart

```diff
@@ -30,7 +33,102 @@ class Console extends StatefulWidget {
 }
 
 class _ConsoleState extends State<Console> {
-  final scrollController = ScrollController();
+  @override
+  Widget build(BuildContext context) {
+    return ValueListenableBuilder<List<String>>(
+      valueListenable: widget.controller.stdio,
+      builder: (context, lines, _) {
+        // Extract this to a reusable ConsoleWithControls StatelessWidget?
+        return Stack(
+          children: [
+            ConsoleOutput(lines: lines),
+            Container(
+              alignment: Alignment.bottomRight,
+              child: ConsoleControls(
+                disabled: lines.isEmpty,
+                onCopyPressed: () {
+                  Clipboard.setData(ClipboardData(
+                    text: lines.join('\n'),
+                  )).then((_) {
+                    Notifications.of(context)?.push(
+                      'Copied!',
+                    );
+                  });
+                },
+                onDeletePressed: widget.controller.clearStdio,
+              ),
+            ),
+          ],
+        );
+      },
+    );
+  }
+}
+
+/// Optionally renders a ButtonBar with Console Controls,
+/// Copy and Clear.
+/// The callbacks for those buttons are passed from the outside.
+class ConsoleControls extends StatelessWidget {
+  const ConsoleControls({
+    this.disabled,
+    this.onCopyPressed,
+    this.onDeletePressed,
+  }) : super();
+
+  final Function onCopyPressed;
```

Comment:
Use `VoidFunction`?

Reply:
I think I'm going to refactor this so the "actions" can be passed from the outside as a `<Widget>[]`, instead of having to pass each callback on its own. That way the user of the Console widget can define what actions are available (for example, in Logging, only "Copy" would be available)